text/3127-trim-paths.md

```diff
@@ -0,0 +1,264 @@
+tune the scope of remapping, specifying paths under which context (in macro expansion, in debuginfo or in diagnostics)
+should or shouldn't be remapped.
+
+# Motivation
+[motivation]: #motivation
+
+## Sanitising local paths that are currently embedded
+Currently, executables and libraries built by Rust and Cargo have a lot of embedded absolute paths. They most frequently appear in debug information and
+panic messages (pointing to the panic location source file). As an example, consider the following package:
+
+`Cargo.toml`:
+
+```toml
+[package]
+name = "rfc"
+version = "0.1.0"
+edition = "2018"
+
+[dependencies]
+rand = "0.8.0"
+```
+`src/main.rs`
+
+```rust
+use rand::prelude::*;
+    
+fn main() {
+    let r: f64 = rand::thread_rng().gen();
+    println!("{}", r);
+}
+```
+
+Then run
+
+```bash
+$ cargo build --release
+$ strings target/release/rfc | grep $HOME
+```
+
+We will see some absolute paths pointing to dependency crates downloaded by Cargo, containing our username:
+
+```
+could not initialize thread_rng: /home/username/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.8.3/src/rngs/thread.rs
+/home/username/.cargo/registry/src/github.com-1ecc6299db9ec823/rand_chacha-0.3.0/src/guts.rsdescription() is deprecated; use Display
+/home/username/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.2/src/util_libc.rs
+```
+
+This is undesirable for the following reasons:
+
+1. **Privacy**. `release` binaries may be distributed, and anyone could then see the builder's local OS account username.
+   Additionally, some CI (such as [GitLab CI](https://docs.gitlab.com/runner/best_practice/#build-directory)) checks out the repo under a path where
+   non-public information is included. Without sanitising the path by default, this may be inadvertently leaked.
+2. **Build reproducibility**. We would like to make it easier to reproduce binary equivalent builds. While it is not required to maintain
+   reproducibility across different environments, removing environment-sensitive information from the build will increase the tolerance on the
+   inevitable environment differences. This helps with build verification, as well as producing deterministic builds when using a distributed build
+   system.
+
+## Handling sysroot paths
+At the moment, paths to the source files of standard and core libraries, even when they are present, always begin with a virtual prefix in the form
+of `/rustc/[SHA1 hash]/library`. This is not an issue when the source files are not present (i.e. when `rust-src` component is not installed), but
+when a user installs `rust-src` they may want the path to their local copy of source files to be visible. Hence the default behaviour when `rust-src`
+is installed should be to use the local path. These local paths should be then affected by path remappings in the usual way.
```

Comment:
The precompiled standard library libraries would use the virtual prefix even if rust-src is installed, so you have to add the remap for debuggers either way.

Reply:
When `rust-src` is installed, rustc internally does a hack to map the virtual prefix to the corresponding real path on the user's local fs: https://github.com/rust-lang/rust/blob/d1a9a9551741c3e888d350d8d4f4821a5addccb2/compiler/rustc_metadata/src/rmeta/decoder.rs#L1477-L1478

Though this is real path is not emitted because it wasn't affected by `--remap-path-prefix` which makes reproducible builds impossible: https://github.com/rust-lang/rust/issues/73167.

However, some people want to see the real path for debugging: https://github.com/rust-lang/rust/issues/85463.

The point here is to satisfy both use cases: without `rust-src` you don't have a choice but to see the virtual path. With `rust-src` you'll see the real path if you don't do anything else, but if you want build reproducibility you can use `--remap-path-prefix` to remap them.

Reply:
i think rust-analyzer requires rust-src. does this mean people using rust-analyzer will not have sanitization by default which i thought would happen by default from other comments

Reply:
Sanitisation by default will happen when you do `cargo build --release` (or `cargo test --release`), regardless of the presence of rust-src. For default debug builds people with rust-src will now see the local paths.

Reply:
ok sounds like i was mixing things up then, sorry.

so regardless of rust-analyzer/rust-src if someone is doing `--release` then its still clean. just reiterating to confirm i have things straight since the discussion was confusing for me.

Reply:
> so regardless of rust-analyzer/rust-src if someone is doing --release then its still clean.

Correct.

Reply:
> The precompiled standard library libraries would use the virtual prefix even if rust-src is installed, so you have to add the remap for debuggers either way.

I think what @bjorn3 is referring to is that the virtual prefixes already embedded in the debuginfo of the standard library, that is, rustc won't have access to them anymore and does not have a way of remapping those paths. Only the linker will touch that debuginfo.